lib/services/local/presence.go

```diff
@@ -1555,6 +1567,110 @@ func (s *PresenceService) ListResources(ctx context.Context, req proto.ListResou
+	var resources []types.ResourceWithLabels
+	switch req.ResourceType {
+	case types.KindNode:
+		nodes, err := s.GetNodes(ctx, req.Namespace)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		servers := types.Servers(nodes)
+		if err := servers.Sort(req.SortBy); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		resources = servers.AsResources()
+
+	case types.KindAppServer:
+		appservers, err := s.GetApplicationServers(ctx, req.Namespace)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		servers := types.AppServers(appservers)
+		if err := servers.Sort(req.SortBy); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		resources = servers.AsResources()
+
+	case types.KindDatabaseServer:
+		dbservers, err := s.GetDatabaseServers(ctx, req.Namespace)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		servers := types.DatabaseServers(dbservers)
+		if err := servers.Sort(req.SortBy); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		resources = servers.AsResources()
+
+	default:
+		return nil, "", trace.NotImplemented("resource type %q is not supported for ListResourcesWithSort", req.ResourceType)
+	}
+
+	return FakePaginate(resources, req)
+}
+
+// FakePaginate is used when we are working with an entire list of resources upfront but still requires pagination.
+func FakePaginate(resources []types.ResourceWithLabels, req proto.ListResourcesRequest) ([]types.ResourceWithLabels, string, error) {
+	limit := int(req.Limit)
+	if limit <= 0 {
+		return nil, "", trace.BadParameter("nonpositive limit value")
+	}
+
+	// Trim resources that precede start key.
+	if req.StartKey != "" {
+		pageStart := 0
+		for i, resource := range resources {
+			cKey := backend.GetPaginationKey(resource)
+			sKey := req.StartKey
+			cLen := len(cKey) - 1
+			sLen := len(sKey) - 1
+
+			// All chars in both key's should match minus the last character.
+			// (The last char of key is what we increment when we call `backend.NextPaginationKey`)
+			// When keys match, we can assume this is the key where we left of.
+			// The last character of the currKey should be less than the startKey.
+			if (cKey[0:cLen] == sKey[0:sLen]) && (cKey[cLen:] < sKey[sLen:]) {
+				pageStart = i + 1
+				break
+			}
+		}
+		resources = resources[pageStart:]
+	}
+
+	// Iterate and filter resources, halting when we reach page limit.
+	var filtered []types.ResourceWithLabels
+	for _, resource := range resources {
+		if len(filtered) == limit {
+			break
+		}
+
+		switch match, err := services.MatchResourceByFilters(resource, services.MatchResourceFilter{
+			ResourceKind:        req.ResourceType,
+			Labels:              req.Labels,
+			SearchKeywords:      req.SearchKeywords,
+			PredicateExpression: req.PredicateExpression,
+		}); {
+		case err != nil:
+			return nil, "", trace.Wrap(err)
+		case match:
+			filtered = append(filtered, resource)
+		}
+	}
+
+	var nextKey string
+	if len(filtered) == limit {
```

Comment:
What happens if we reach the limit right as we reach the end of the resources? Are we ok with sending a `nextKey` that will result in an empty fetch for the next request? Is this already done in other places that do pagination?

Reply:
> Are we ok with sending a nextKey that will result in an empty fetch for the next request? Is this already done in other places that do pagination?

yes

i double checked with forrest that this was fine (this func was a copy of [listNodesFromTTLCache](https://github.com/gravitational/teleport/blob/master/lib/cache/cache.go#L1355) with some modifications). We can't accurately tell if there will be a nextKey unless we filter for every row (which we do not, we stop filter when we reach limit). So it was faster to fetch just to return empty versus running the filter for every row just to know if we have a next key.

Reply:
But that'd only work if we're actually at the end of all resources, right? Not just the matching resources.

I didn't think of the fact that `resources` is getting filtered in the loop, it's probably better like this for consistency.